Deprecate `set_system_attr` in Study and Trial

cross32768 opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation
As in [document of `set_system_attr`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.set_system_attr), this method is mainly used internally.
Therefore, I propose to hide from normal users.
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->

## Description of the changes
* Change method name `set_system_attr` -> `_set_system_attr` in Study and Trial.
<!-- Describe the changes in this PR. -->

c-bata commented on optuna/integration/_lightgbm_tuner/optimize.py:
```diff
@@ -272,10 +272,10 @@ def _postprocess(
-        trial.set_system_attr(_AVERAGE_ITERATION_TIME_KEY, average_iteration_time)
-        trial.set_system_attr(_STEP_NAME_KEY, self.step_name)
-        trial.set_system_attr(_LGBM_PARAMS_KEY, json.dumps(self.lgbm_params))
+        trial._set_system_attr(_ELAPSED_SECS_KEY, elapsed_secs)
+        trial._set_system_attr(_AVERAGE_ITERATION_TIME_KEY, average_iteration_time)
+        trial._set_system_attr(_STEP_NAME_KEY, self.step_name)
+        trial._set_system_attr(_LGBM_PARAMS_KEY, json.dumps(self.lgbm_params))
```
Could you call `storage.set_trial_system_attrs` instead?

c-bata replied:
I'd say we can get the `storage` object via `trial._storage`. Although it's a private attribute, it's acceptable to use inside the optuna's implementation.

cross32768 replied:
Thank you for you advice! (and sorry for late reply)

I update my code to use `trial._storage.set_trial_system_attrs(trial._trial_id, key, value)` instead of `trial.set_system_attrs(key, value)`.

I also use `study._storage.set_study_system_attrs(study._study_id, key, value)` instead of `study.set_system_attrs(key, value)`.

c-bata commented on optuna/trial/_trial.py:
```diff
@@ -557,7 +557,7 @@ def objective(trial):
         self.storage.set_trial_user_attr(self._trial_id, key, value)
 
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def _set_system_attr(self, key: str, value: Any) -> None:
```
Could you avoid renaming `set_system_attr` to `_set_system_attr` since it's a breaking change. We should just deprecate it (by wrapping with a `@deprecated` decorator) and then update the docstring.

cross32768 replied:
I added @deprecated_func("3.1.0", "6.0.0") to `set_system_attr` method.

cross32768 replied:
For now, I set the timing of deprecation `"6.0.0"`. Is it reasonable choice? Or too far?

c-bata replied:
Sounds reasonable 👍 It follows our deprecation policy.
https://github.com/optuna/optuna/wiki/Deprecation-policy

not522 replied:
By default, we will remove it in v5.0.0 (two major versions after v3.1.0). But I agree to defer the version because users seem to use it, and the maintenance cost is low.

c-bata commented on tests/study_tests/test_study.py:
```diff
@@ -300,7 +300,7 @@ def test_study_set_and_get_system_attrs(storage_mode: str) -> None:
         study = create_study(storage=storage)
 
-        study.set_system_attr("system_message", "test")
+        study._set_system_attr("system_message", "test")
```
It might be a bit controversial but I'd say we can just remove test scenarios that checks the behavior of `set_system_attr`. What do you think?

cross32768 replied:
Will we continue to use `set_system_attr` internally for the future? If so, I think it's better to test.

c-bata replied:
No, we will internally use `storage.set_trial_system_attr` instead of `trial.se_system_attr`. However, as you said, it might be better to leave tests for `trial.set_system_attr`.

@gen740 (cc: @HideakiImamura ) What do you think?

cross32768 replied:
I heard that deprecated function tests are also deprecated customarily, so I deleted them.

c-bata reviewed (commented):
@cross32768  Thank you for your pull request!
I understand it's WIP but left some suggestions since I noticed while quickly taking a look.

cross32768 commented:
I wil add document about deprecation to Docstrings later.

HideakiImamura commented:
@gen740 Could you review this PR?

HideakiImamura commented:
Let me convert this PR to draft. 

@cross32768 If you make this PR passing CI and remaining TODOs, please convert this ready for review.

cross32768 commented:
* Change point
  * I replaced `trial.set_system_attrs(key, value)` by `trial.storage.set_trial_system_attrs(trial._trial_id, key, value)`
  * I also replaced `study.set_system_attrs(key, value)` by `study._storage.set_study_system_attrs(study._study_id, key, value)`
  * I removed some tests about `system_attrs`

cross32768 commented:
* Current problem & discusion
  * Current code cannot pass doc test. But I cannot read which change is the problem from the error. What should I do?
  * `FrozenTrial` class does not have `storage`. Thus, I cannot replace `FrozenTrial.set_system_attr` with other way. 
  * Should I add the test for correct deprecation warning?
  * Should I also make`set_system_attrs` deprecated? This change make big effect to a lot of files (over 50 files).

c-bata commented on optuna/multi_objective/study.py:
```diff
@@ -373,6 +375,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
 
         self._study.set_user_attr(key, value)
 
+    @deprecated_func("3.1.0", "6.0.0")
```
nit: We can omit the deprecation message here since multi_objective module itself is deprecated.
```suggestion
```

c-bata commented on optuna/multi_objective/trial.py:
```diff
@@ -162,14 +163,15 @@ def set_user_attr(self, key: str, value: Any) -> None:
 
         self._trial.set_user_attr(key, value)
 
+    @deprecated_func("3.1.0", "6.0.0")
```
ditto

```suggestion
```

c-bata reviewed (commented):
LGTM with nits.

c-bata commented:
Let me change the reviewers from @gen740 to @HideakiImamura since Fujimoto-san is busy this week.

c-bata commented:
@not522 Could you review this PR?

not522 reviewed (approved):
Thank you for your PR. LGTM.

abraaoz commented:
I use this method as follows:
```py
trial_note: str = "markdown here"
study.set_system_attr(f"dashboard:{trial.number}:note_ver", 1)
study.set_system_attr(f"dashboard:{trial.number}:note_str:0", trial_note)
```
How can I achieve the same result without using `set_system_attr`?
User attributes don’t appear as markdown in Optuna Dashboard.

c-bata commented:
@abraaoz You can use the `optuna_dashboard.save_note()` function.
https://optuna-dashboard.readthedocs.io/en/latest/_generated/optuna_dashboard.save_note.html#optuna_dashboard.save_note

For further questions, please use [GitHub Discussions](https://github.com/optuna/optuna/discussions/categories/q-a) instead.
